android/src/main/java/io/ably/flutter/plugin/push/PushMessagingEventHandlers.java

```diff
@@ -16,21 +15,52 @@
 import io.ably.flutter.plugin.generated.PlatformConstants;
 import io.flutter.plugin.common.MethodChannel;
 
+/**
+ * This class is used when the application is in the **foreground** or **background**, but not when
+ * **terminated**. See [PushTerminatedIsolateRunner.java] to see where push notifications being
+ * handled whilst the app is terminated.
+ */
 final public class PushMessagingEventHandlers {
   private static final String TAG = PushMessagingEventHandlers.class.getName();
   public static final String PUSH_ON_MESSAGE_RECEIVED = "io.ably.ably_flutter.PUSH_ON_MESSAGE_RECEIVED";
   public static final String PUSH_ON_BACKGROUND_MESSAGE_RECEIVED = "io.ably.ably_flutter.PUSH_ON_BACKGROUND_MESSAGE_RECEIVED";
   public static final String PUSH_ON_BACKGROUND_MESSAGE_PROCESSING_COMPLETE = "io.ably.ably_flutter.PUSH_ON_BACKGROUND_MESSAGE_COMPLETE";
   static PushMessagingEventHandlers instance;
 
-  public static void instantiate(Context context, MethodChannel methodChannel) {
-    if (instance == null) {
-      instance = new PushMessagingEventHandlers(context, methodChannel);
+  /**
+   * Replaces the existing instance if required. this is because the latest methodChannel is the
+   * most important because it is from the latest plugin registration.
+   * @param context
+   * @param methodChannel
+   */
+  public static void reset(Context context, MethodChannel methodChannel) {
+    if (instance != null) {
+      instance.close(context);
     }
+    instance = new PushMessagingEventHandlers(context, methodChannel);
   }
 
+  private void close(Context context) {
+    this.broadcastReceiver.close(context);
+  }
+
+  private final BroadcastReceiver broadcastReceiver;
   private PushMessagingEventHandlers(Context context, MethodChannel methodChannel) {
-    new BroadcastReceiver(context, methodChannel);
+    this.broadcastReceiver = new BroadcastReceiver(context, methodChannel);
+  }
+
+  // Send message to Dart side app already running
+  public static void sendMessageToFlutterApp(Context context, Intent intent) {
+    final Intent onMessageReceivedIntent = new Intent(PUSH_ON_MESSAGE_RECEIVED);
+    onMessageReceivedIntent.putExtras(intent.getExtras());
+    LocalBroadcastManager.getInstance(context).sendBroadcast(onMessageReceivedIntent);
+  }
+
+  // Flutter is already running, just send a background message to it.
+  public static void sendBackgroundMessageToFlutterApp(Context context, Intent intent) {
```

Comment:
NonNull annotation...

Reply:
I would say please assume parameters are non null unless they are declared to be nullable. However, I have also added them here: 3021a28d0802b54ca5bbf7469122acbf99b6874f